src/client/java/teammates/client/scripts/sql/DataMigrationForCourseEntitySql.java

```diff
@@ -148,4 +186,200 @@ private Student createStudent(teammates.storage.sqlentity.Course newCourse,
 
         return newStudent;
     }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    protected void doOperation() {
+        log("Running " + getClass().getSimpleName() + "...");
+        log("Preview: " + isPreview());
+        setMigrationCriteria();
+
+        Cursor cursor = readPositionOfCursorFromFile().orElse(null);
+        if (cursor == null) {
+            log("Start from the beginning");
+        } else {
+            log("Start from cursor position: " + cursor.toUrlSafe());
+        }
+
+        boolean shouldContinue = true;
+        while (shouldContinue) {
+            shouldContinue = false;
+            Query<Course> filterQueryKeys = getFilterQuery().limit(BATCH_SIZE);
+            if (cursor != null) {
+                filterQueryKeys = filterQueryKeys.startAt(cursor);
+            }
+
+            QueryResults<?> iterator = filterQueryKeys.iterator();
+
+            Course currentCourse = null;
+            // Cascade delete the course if it is not fully migrated.
+            if (iterator.hasNext()) {
+                currentCourse = (Course) iterator.next();
+                if (currentCourse.isMigrated()) {
+                    currentCourse = (Course) iterator.next();
+                } else {
+                    deleteCourseCascade(currentCourse);
+                }
+            }
+
+            while (currentCourse != null) {
+                shouldContinue = true;
+                migrateWithoutTrx(currentCourse);
+                numberOfScannedKey.incrementAndGet();
+
+                cursor = iterator.getCursorAfter();
+                savePositionOfCursorToFile(cursor);
+
+                currentCourse = iterator.hasNext() ? (Course) iterator.next() : null;
+            }
+
+            if (shouldContinue) {
+                log(String.format("Cursor Position: %s", cursor.toUrlSafe()));
+                log(String.format("Number Of Course Entity Key Scanned: %d", numberOfScannedKey.get()));
+                log(String.format("Number Of Course Entity affected: %d", numberOfAffectedEntities.get()));
+                log(String.format("Number Of Course Entity updated: %d", numberOfUpdatedEntities.get()));
+            }
+        }
+
+        deleteCursorPositionFile();
+        log(isPreview() ? "Preview Completed!" : "Migration Completed!");
+        log("Total number of course entities: " + numberOfScannedKey.get());
+        log("Number of affected course entities: " + numberOfAffectedEntities.get());
+        log("Number of updated course entities: " + numberOfUpdatedEntities.get());
+    }
+
+    /**
+     * Deletes the course and its related entities from sql database.
+     */
+    private void deleteCourseCascade(Course oldCourse) {
+        // TODO: Implement deleteCourseCascase
+        log("delete course id: " + oldCourse.getUniqueId());
```

Comment:
I think this is a one-liner, could we do this in this PR?

Reply:
I don't think calling HibernateUtil.remove(course) can cascade delete other entities, maybe need to do something like in [CourseLogic](https://github.com/TEAMMATES/teammates/blob/master/src/main/java/teammates/sqllogic/core/CoursesLogic.java#L119)

Reply:
I think you're right, I was originally thinking of using [Cascade on delete annotation](https://docs.jboss.org/hibernate/orm/current/javadocs/org/hibernate/annotations/OnDelete.html) but there may be performance issues with this